core/src/main/java/org/infinispan/security/impl/AuthorizationHelper.java

```diff
@@ -88,44 +93,60 @@ private void checkSecurityManagerPermission(AuthorizationPermission perm) {
       }
    }
 
-   public int computeSubjectRoleMask(Subject subject, AuthorizationConfiguration configuration) {
+   private boolean checkSubjectPermissionAndRole(Subject subject, AuthorizationConfiguration configuration,
+         AuthorizationPermission requiredPermission, String requestedRole) {
       if (subject != null) {
-         Integer cachedMask;
-         try {
-            cachedMask = maskCache != null ? maskCache.get(maskCacheScope, subject) : null;
-         } catch (IllegalStateException e) {
-            cachedMask = null;
+         CachePrincipalPair cpp = new CachePrincipalPair(getUserPrincipal(subject), name);
+         SubjectACL subjectACL;
+         if (aclCache != null)
+            subjectACL = aclCache.computeIfAbsent(cpp, s -> {
+               return computeSubjectACL(subject, configuration);
+            });
+         else
+            subjectACL = computeSubjectACL(subject, configuration);
+
+         int permissionMask = requiredPermission.getMask();
+         return subjectACL.matches(permissionMask) && (requestedRole != null ? subjectACL.containsRole(requestedRole) : true);
+      } else {
+         return false;
+      }
+   }
+
+   private SubjectACL computeSubjectACL(Subject subject, AuthorizationConfiguration configuration) {
+      PrincipalRoleMapper roleMapper = globalConfiguration.authorization().principalRoleMapper();
+      Set<Principal> principals = subject.getPrincipals();
+      Set<String> allRoles = new HashSet<>(principals.size());
```

Comment:
I would say that maybe not all principals would map to a role. The case where a principal might map to multiple roles is unlikely in my experience.